staging/src/k8s.io/api/autoscaling/v2/types.go

```diff
@@ -0,0 +1,536 @@
+	// +optional
+	Pods *PodsMetricSource `json:"pods,omitempty" protobuf:"bytes,3,opt,name=pods"`
+	// resource refers to a resource metric (such as those specified in
+	// requests and limits) known to Kubernetes describing each pod in the
+	// current scale target (e.g. CPU or memory). Such metrics are built in to
+	// Kubernetes, and have special scaling options on top of those available
+	// to normal per-pod metrics using the "pods" source.
+	// +optional
+	Resource *PodResourceMetricSource `json:"resource,omitempty" protobuf:"bytes,4,opt,name=resource"`
+	// containerResource refers to a resource metric (such as those specified in
+	// requests and limits) known to Kubernetes describing a single container in
+	// each pod of the current scale target (e.g. CPU or memory). Such metrics are
+	// built in to Kubernetes, and have special scaling options on top of those
+	// available to normal per-pod metrics using the "pods" source.
+	// This is an alpha feature and can be enabled by the HPAContainerMetrics feature flag.
+	// +optional
+	ContainerResource *ContainerResourceMetricSource `json:"containerResource,omitempty" protobuf:"bytes,7,opt,name=containerResource"`
+	// external refers to a global metric that is not associated
+	// with any Kubernetes object. It allows autoscaling based on information
+	// coming from components running outside of cluster
+	// (for example length of queue in cloud messaging service, or
+	// QPS from loadbalancer running outside of cluster).
+	// +optional
+	External *ExternalMetricSource `json:"external,omitempty" protobuf:"bytes,5,opt,name=external"`
+}
+
+// HorizontalPodAutoscalerBehavior configures the scaling behavior of the target
+// in both Up and Down directions (scaleUp and scaleDown fields respectively).
+type HorizontalPodAutoscalerBehavior struct {
+	// scaleUp is scaling policy for scaling Up.
+	// If not set, the default value is the higher of:
+	//   * increase no more than 4 pods per 60 seconds
+	//   * double the number of pods per 60 seconds
+	// No stabilization is used.
+	// +optional
+	ScaleUp *HPAScalingRules `json:"scaleUp,omitempty" protobuf:"bytes,1,opt,name=scaleUp"`
+	// scaleDown is scaling policy for scaling Down.
+	// If not set, the default value is to allow to scale down to minReplicas pods, with a
+	// 300 second stabilization window (i.e., the highest recommendation for
+	// the last 300sec is used).
+	// +optional
+	ScaleDown *HPAScalingRules `json:"scaleDown,omitempty" protobuf:"bytes,2,opt,name=scaleDown"`
+}
+
+// ScalingPolicySelect is used to specify which policy should be used while scaling in a certain direction
+type ScalingPolicySelect string
+
+const (
+	// MaxChangePolicySelect  selects the policy with the highest possible change.
+	MaxChangePolicySelect ScalingPolicySelect = "MaxChange"
+	// MinChangePolicySelect selects the policy with the lowest possible change.
+	MinChangePolicySelect ScalingPolicySelect = "MinChange"
```

Comment:
I changed ResourceMetricSourceType  to PodResourceMetricSourceType and Resource to PodResource as following with #102362. But it is not very clear to me why we should  rename ResourceMetricSourceType to PodResourceMetricSourceType . @josephburnett Could you tell @liggitt   why we should rename ResourceMetricSourceType to PodResourceMetricSourceType? Thanks.

Reply:
I will try to  recover MaxChangePolicySelect to MaxPolicySelect and MaxChange to Max

Reply:
@liggitt yes, I think it's worth the extra clarity.  I've had quite a few from GKE users about what these mean, and they usually get it wrong for scale down controls.  

E.g. policies of 2x and 4 pods when scaling down from 100 will allow 50 and 96 respectively.  Using select policy "Max" will allow you to scale down to 50 (not 96 which is a larger value).  That's because "Max" means selecting the policy that allows the largest change.

This is one of the things we've learned from the v2 API being in beta.

Reply:
anything can be done with enough effort, but the rename will make our lives harder since v2 conversions will have to be manually written and validation will have to be version-specific, and will make existing users' lives harder, since the v2beta2 manifests will not be valid v2 manifests

Reply:
Yes, we should try and maintain compatibility with existing manifests.  However adoption of the `behavior` struct is low (so far) and those customers that have used it were often confused by these values.  In this particular case I believe we will induce less confusion overall by fixing it now.